take 2 of ibis 3.x + substrait + duckdb

gforsyth opened the pull request:
Minor annoyance:
The syntax for `strftime` in SQLite is `strftime(format string, date)`
but in DuckDB it is `strftime(date, format string)` so while the sets of
sql queries are _nearly_ identical I opted to create a separate
directory for the duckdb ones

Using a barebones backend for handling the unbound tables for substrait serialization.

Using `dateutil` to fix some bugs in the date changer

Also, holy cow how was `order_by` in query 1 not breaking things?

---

Info from #2 

This got bigger than I expected.  It does a few things:

1. Removes all usage of `materialize` (with eyes towards the release of Ibis 3.x)
2. Adds one level of abstraction around the Ibis query functions so that it is possible to pass in a set of tables directly, instead of a connection.  This allows for creating and passing in unbound Ibis tables with an eye towards...
3. Adding in ibis-substrait to compile and then serialize the queries to protobuf, before then...
4. Attempting to deserialize and execute them using DuckDB.

Currently, all 22 queries are failing :raised_eyebrow

Two of the failure modes in DuckDB look like they probably aren't too hard to fix?  Those are the "single grouping sets" not supported error, and the Catalog Errors, which look like they probably just need a lookup table?

I haven't yet dug in to the nature of the `INTERNAL Error` messages.

| Query | Error                                                                                                                              | Substrait or DuckDB |
|-------|------------------------------------------------------------------------------------------------------------------------------------|---------------------|
| h01   | RuntimeError: INTERNAL Error: Only single grouping sets are supported for now                                                      | DuckDB              |
| h02   | TypeError: Parameter to MergeFrom() must be instance of same class: expected substrait.Expression got substrait.AggregateFunction. | Substrait           |
| h03   | RuntimeError: INTERNAL Error: Only single grouping sets are supported for now                                                      | DuckDB              |
| h04   | RuntimeError: INTERNAL Error: 16                                                                                                   | DuckDB              |
| h05   | RuntimeError: INTERNAL Error: 2                                                                                                    | DuckDB              |
| h06   | RuntimeError: INTERNAL Error: 16                                                                                                   | DuckDB              |
| h07   | NotImplementedError: (SelfReference(UnboundTable\[r0, name=nation\]                                                                | Substrait           |
| h08   | NotImplementedError: (SelfReference(UnboundTable\[r0, name=nation\]                                                                | Substrait           |
| h09   | RuntimeError: INTERNAL Error: Only single grouping sets are supported for now                                                      | DuckDB              |
| h10   | RuntimeError: INTERNAL Error: Only single grouping sets are supported for now                                                      | DuckDB              |
| h11   | RuntimeError: Catalog Error: Scalar Function with name equals does not exist!                                                      | DuckDB              |
| h12   | RuntimeError: INTERNAL Error: 2                                                                                                    | DuckDB              |
| h13   | RuntimeError: Catalog Error: Scalar Function with name equals does not exist!                                                      | DuckDB              |
| h14   | TypeError: Parameter to MergeFrom() must be instance of same class: expected substrait.Expression got substrait.AggregateFunction. | Substrait           |
| h15   | AssertionError: non-empty child<sub>relfieldoffsets</sub> passed in to selection translation rule                                  | Substrait           |
| h16   | RuntimeError: INTERNAL Error: Only single grouping sets are supported for now                                                      | DuckDB              |
| h17   | TypeError: Parameter to MergeFrom() must be instance of same class: expected substrait.Expression got substrait.AggregateFunction. | Substrait           |
| h18   | RuntimeError: INTERNAL Error: Only single grouping sets are supported for now                                                      | DuckDB              |
| h19   | RuntimeError: INTERNAL Error: 2                                                                                                    | DuckDB              |
| h20   | RuntimeError: Catalog Error: Scalar Function with name equals does not exist!                                                      | DuckDB              |
| h21   | RuntimeError: Catalog Error: Scalar Function with name equals does not exist!                                                      | DuckDB              |
| h22   | RuntimeError: INTERNAL Error: 2                                                                                                    | DuckDB              |

(also this currently needs to run against a _very_ fresh DuckDB master since the `get_substrait` and `from_substrait` PR was only merged on Friday 2022/03/18

gforsyth committed:
test: add tests for ibis -> substrait -> duckdb TPC-H queries

All xfailed for now

gforsyth committed:
test: add in dummy backend to handle passing unbound tables

gforsyth committed:
test: add tests for ibis -> substrait -> duckdb TPC-H

fix: use dateutil for relative time deltas

fix: use add_date helper and fix order_by -> sort_by

test: change all xfails to expected error types

ci: cache pip and run substrait serde tests

refactor: remove unused pytest markers

chore: add pytest to gha

chore: remove unused pytest opts

gforsyth commented:
Also https://github.com/ibis-project/ibis/commit/1f2b3fae8f8259da64004e1efc3849c59939081a broke query 4

Mytherin commented:
I have pushed a fix for the `strftime` argument order in duckdb/duckdb#3295 which should allow the SQLite order to be used in DuckDB as well. I would recommend against using `strftime` in these queries and use [`EXTRACT`](https://modern-sql.com/feature/extract) instead, however. That seems like a work-around used for the SQLite version of the queries because SQLite does not support the SQL-standard `EXTRACT` command. `EXTRACT` will be faster since it will avoid printing results to a string unnecessarily.

cpcloud commented:
@Mytherin Thank you for the fix.

Ideally we could use extract, but since we're comparing the outputs across all these systems the simplest thing to do initially was `strftime`. We could probably use extract, and then in the SQLite queries add a `CAST` to `INT`, but I'm not sure if at the moment we're uber concerned about performance comparisons (yet!). Anyway, really appreciate the fix and the extract tip.

saulpw commented on ibis_tpc/h19.py:
```diff
@@ -20,32 +20,32 @@ def tpc_h19(
 
     q1 = (
         (q.p_brand == BRAND1)
-        & (q.p_container.isin(("sm case", "sm box", "sm pack", "sm pkg")))
```
How were these working before?  (or weren't they?)

gforsyth replied:
Apparently `sqlite` doesn't care about case at all?

gforsyth replied:
hmm, no wait, that's not right.  I have no idea how it was working

saulpw commented on ibis_tpc/h14.py:
```diff
@@ -14,7 +14,7 @@ def tpc_h14(con, DATE="1995-09-01"):
     q = q.filter([q.l_shipdate >= DATE, q.l_shipdate < add_date(DATE, dm=1)])
 
     revenue = q.l_extendedprice * (1 - q.l_discount)
-    promo_revenue = q.p_type.like("promo%").ifelse(revenue, 0)
```
is there an `ilike()` or another way to do case-insensitive LIKE?

cpcloud replied:
Yep [`ilike`](https://ibis-project.org/api/expressions/strings/#ibis.expr.types.strings.StringValue.ilike) is a thing!

gforsyth replied:
I guess the question is how closely we want to hew to the TPC-H queries as they are defined

saulpw commented on ibis_tpc/runners.py:
```diff
@@ -278,29 +278,72 @@ def compare(rows1, rows2):
+@dispatch(pandas.Timestamp, pandas.Timestamp)
+def _compare(v1, v2, row=None, key=None, diffs=None):
+    if v1 != v2:
+        diffs.append(f"[{row}].{key} (date) {v1} != {v2}")
```
How about `return` or `yield` instead of passing in the container itself?

gforsyth replied:
Yep, that'll be much cleaner

saulpw reviewed (commented):
Thanks for multipledispatch changes, that looks so much nicer.

gforsyth committed:
refactor: update parse_type import

gforsyth committed:
feat: add substrait binary blobs

cpcloud commented:
The SQLAlchemy version needs to be bumped for the `master` build, just noting here before release.

gforsyth committed:
test: make substrait tests fail correctly

I overengineered this originally -- now tests pass if the query compiles
/ executes and fails if not.

The strict xfails should alert us to when things are working as expected

gforsyth committed:
test: simplify ci jobs a bit

Pull `ibis-substrait` from pypi since it gets released frequently.

And no need to run against ibis-master anymore since we run tpc-queries
in the ibis CI.

gforsyth committed:
feat: add substrait runner

Currently it only runs on `duckdb` -- we can make it extensible later.

Other small changes:
* A runner returning empty `rows` now raises an error since all queries
  should return _some_ output
* Moving some substrait helper stuff into a module instead of `conftest.py`

gforsyth committed:
ci: add substrait runs to CI (this will fail)

gforsyth committed:
chore: regenerate tpc-h substrait blobs and add helper script

gforsyth committed:
test: add roundtrip compile/decompile tests

gforsyth commented:
Currently having deserialization issues on the duckdb side due to (I think) missing mappings for the following scalar functions:

- `add`
- `substract`
- `multiply`
- `divide`
- `cast`

and missing functionality to deserialize `SingularOrList` which corresponds to our `ops.Contains` (https://github.com/substrait-io/substrait/blob/main/proto/substrait/algebra.proto#L521-L524=)

hannes commented:
CC @pdet

pdet commented:
@gforsyth, what's the difference between the ```cast``` function and https://github.com/substrait-io/substrait/blob/main/proto/substrait/algebra.proto#L498-L508= ?

gforsyth commented:
> @gforsyth, what's the difference between the `cast` function and [https://github.com/substrait-io/substrait/blob/main/proto/substrait/algebra.proto#L498-L508=](https://github.com/substrait-io/substrait/blob/main/proto/substrait/algebra.proto?rgh-link-date=2022-05-19T13%3A22%3A59Z#L498-L508=) ?

whoops -- my bad -- I just put it in the wrong list.  We should be using the `Cast` expression that you linked to.

pdet commented:
@gforsyth can you also send me a binary that does the ```SingularOrList``` for testing?

Our in-clauses get turned into mark joins, so we don't output these ourselves.

gforsyth committed:
test: update xfail reasons, don't randomize test order

gforsyth commented:
> @gforsyth can you also send me a binary that does the `SingularOrList` for testing?

Sure, you can grab the blob for query 12 which has one: https://github.com/gforsyth/tpc-queries/blob/tpc_substrait2/substrait_tpc/tpc_h12.bin

it's the bit that corresponds to the check on the value of `l_shipmode` being either `"MAIL"` or `"SHIP"`
```proto
args {
  singular_or_list {
    value {
      selection {
        direct_reference {
          struct_field {
            field: 23
          }
        }
        root_reference {
        }
      }
    }
    options {
      literal {
        string: "MAIL"
      }
    }
    options {
      literal {
        string: "SHIP"
      }
    }
  }
}
```

gforsyth commented:
@pdet -- also, if you want to use this branch for testing, the following will run all 22 TPC-H queries through `ibis-substrait` and then attempt to execute them on `duckdb` via `from_substrait`:

```
pytest -vv -k send_to_duckdb --runxfail
```

(that should pass through the failure messages -- a bunch of those are on our end but you'll be able to see which queries trigger `INTERNAL Error` on the duckdb side)

Happy to hop on a call to get an environment spun up to run the above if that's helpful.

pdet commented:
Jup, is there any doc on how to get the ibis_substrait going?
Otherwise, I would surely appreciate help on getting things going; I guess having some kind of round-trip test env is way more useful than the constant binary sharing :-)

gforsyth commented:
> Jup, is there any doc on how to get the ibis_substrait going?

you should be able to `pip install ibis_substrait` and everything in CI here is running against the latest release (and we cut a release for every PR right now)

pdet commented:
Hey @gforsyth , 
I've integrated Ibis to DuckDB's CI.
We have 7 queries passing and 15 broken.
Here is the summary of the current failing TPC-H queries on my branch:

Ibis can't compile query to substrait:
[2,7,8,14,15,16,17,18,20]

DuckDB is not supporting subquery ingestion (Scalar Function any, not...)
[4,13,21,22]

Scalar Function with name extractyear does not exist!
[(function should be named 'extract' with option YEAR) ](https://github.com/substrait-io/substrait/blob/018da380bb91ab4825b205d6f5bb74f5cc435a43/extensions/functions_datetime.yaml#L5)
[9]

Result not matching
[11]

I think [9] should be an easy fix (at least for this current problem). We are not following the standardization of the extract year scalar function. As soon as I figure out why 11 is giving wrong results, I'll implement the consumer-side of the subqueries.

On a different note, can you point me out to the code where I can make Ibis consume substrait plans?

Cheers,

gforsyth commented:
Hey @pdet -- sorry for the delay, I was on vacation the last two weeks.  [9] does look like and easy fix, so that's good!  And we should have subquery compilation working before too long.

If you want to consume substrait plans with ibis, you can do the following:

```python
import ibis
from ibis_substrait.compiler.decompile import decompile

con = ibis.duckdb.connect("tpch.ddb")

(expr,) = decompile(proto_byte_string)

con.execute(expr)
```

gforsyth committed:
test: update failure reasons

gforsyth committed:
chore: manually load substrait extension

gforsyth committed:
chore: update substrait blobs

gforsyth committed:
test: ignore failures in substrait runner (for now)

Note that the strict xfail on the corresponding `pytest` stuff _will_
start yelling if tests start to pass

gforsyth committed:
test: update test failure reasons and raises

gforsyth commented:
Hey @saulpw -- I think this is ready for review.  Some stuff isn't working, but that should change as we fix up ibis-substrait.

saulpw commented:
Thanks @gforsyth !  It looks fine to me, it's now merged.
